### PR TITLE
Callbacks: 1

### PR DESCRIPTION
**Pull request by akshay1y2**

### CODE REVIEW

File: app/models/product.rb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 class Product < ApplicationRecord
 	has_many :line_items
 	has_many :orders, through: :line_items
+	
+	after_initialize { |p| p.title = 'abc' unless p.title }
```
Comment (siddharth28) [inline app/models/product.rb]: @akshay1y2 why is this still in proc?

File: app/models/product.rb [hunk 2]
```diff
@@ -11,16 +13,13 @@ class Product < ApplicationRecord
...
 	validates :price, numericality: { greater_than_or_equal_to: 0.01 }, allow_blank: true, price: true
 
 	private
 
-	def discount_without_price
-		if discount_price && !price
-			errors.add(:discount_price, 'without price mentioned')
-			throw :abort
-		end
+	def set_discount_price
```
Comment (siddharth28) [inline app/models/product.rb]: @akshay1y2 please follow 2 space indentation.

File: app/models/user.rb [hunk 3]
```diff
@@ -1,16 +1,33 @@
 class User < ApplicationRecord
+  ADMIN_EMAIL_REGEX = /\Aadmin@depot.com\z/i
```
Comment (siddharth28) [inline app/models/user.rb]: @akshay1y2 why are we using a regex for admin email?

File: app/models/user.rb [hunk 4]
```diff
@@ -1,16 +1,33 @@
...
+
   validates :name, presence: true, uniqueness: true
   validates :email, uniqueness: true, format: { with: URI::MailTo::EMAIL_REGEXP } 
   has_secure_password
-  after_destroy :ensure_an_admin_remains
+
+  after_commit :send_welcome_mail, on: :create
+  after_save :ensure_admin_is_not_updated
```
Comment (siddharth28) [inline app/models/user.rb]: @akshay1y2 this is not after logic.

File: app/models/user.rb [hunk 5]
```diff
@@ -1,16 +1,33 @@
...
+  before_destroy :ensure_an_admin_remains, :ensure_not_an_admin
```
Comment (siddharth28) [inline app/models/user.rb]: @akshay1y2 ensure_admin_remains is not before logic.

File: app/models/product.rb [hunk 6]
```diff
@@ -1,25 +1,35 @@
...
   validates :price, numericality: { greater_than_or_equal_to: 0.01 }, allow_blank: true, price: true
-  # validates :price, numericality: { greater_than_or_equal_to: :discount_price }, if: :discount_price?, allow_blank: true
 
   private
-  # ensure that there are no line items referencing this product
-  def ensure_not_referenced_by_any_line_item
-    unless line_items.empty?
-      errors.add(:base, 'Line Items present')
-      throw :abort
+    def set_title_to_abc
```
Comment (siddharth28) [inline app/models/product.rb]: @akshay1y2 incorrect method name.

File: app/models/user.rb [hunk 7]
```diff
@@ -1,16 +1,32 @@
...
   class Error < StandardError
   end
 
   private
+    def send_welcome_mail
+      UserMailer.with(user: self).welcome.deliver_later
+    end
+
+    def ensure_admin_is_not_updated
+      raise Error.new "Can't update admin" if email == ADMIN_EMAIL || email_was == ADMIN_EMAIL
```
Comment (siddharth28) [inline app/models/user.rb]: @akshay1y2 why are we raising error?
Add an error and throw abort here.

File: app/models/user.rb [hunk 8]
```diff
@@ -1,16 +1,32 @@
...
+    end
+
+    def ensure_not_an_admin
```
Comment (siddharth28) [inline app/models/user.rb]: @akshay1y2 why are we raising error?
Add an error and throw abort here.

**Comment by akshay1y2** [comment]
https://docs.google.com/document/d/17GD0udHhXzTmRNOZZy_nsysIBnafIomr_xi4OkL-jYQ/edit?pli=1#

**Pull request merged**